feat: more versatile `class abbrev`

tydeu opened the pull request:
Addresses #436. The `class abbrev` command now supports implicitly bound parameters (predefined or otherwise). It also now supports a proper `declId` as opposed to a simple `ident`. It still does not support explicit predefined parameters. I have also added tests for these conditions. I also refactored some of the more useful pieces of the macro's logic into helper functions so that they can be used by end-users (and I can vouch that they would be useful to me at least).

leodemoura commented on src/Init/NotationExtra.lean:
```diff
@@ -140,41 +140,48 @@ macro_rules
+
+namespace Lean.Syntax
+
+def paramsToVars (params : Array Syntax) : MacroM (Array Syntax × Array Syntax) := do
```
Please add comments explaining the macro.

leodemoura commented on tests/lean/classAbbrev.lean:
```diff
@@ -0,0 +1,20 @@
+variable {A : Sort u}
+class Bar (A : Sort u) (f : A -> Prop) := (bar : (a : A) -> f a)
+class abbrev BarAbbrev (f : A -> Prop) := Bar A f
+end
```
Please add move tests.

- You are not using the abbreviations.
- You don't have an abbreviation where the RHS has more than one element.

leodemoura commented on src/Init/NotationExtra.lean:
```diff
@@ -140,41 +140,48 @@ macro_rules
-      auxBinders := auxBinders.push (← `(bracketedBinder| [ $parent:term ]))
-      ctorArgs   := ctorArgs.push inferInst
+  | `(class abbrev $id $params* := $[ $parents:term $[,]? ]*) => do
+    let (name, uvars) := Syntax.expandDeclId id
```
You don't have a test where you provide the universe variables at the `declId`.

leodemoura commented on tests/lean/classAbbrev.lean.expected.out:
```diff
@@ -0,0 +1,3 @@
+classAbbrev.lean:4:13-4:22: error: type expected
```
When a test is supposed to fail, please add a comment on the `.lean` explaining why.

leodemoura commented:
>  It still does not support explicit predefined parameters. 

What is the problem here?

tydeu commented:
@leodemoura Thanks for all the feedback! I will work on addressing them. 

For the case of explicit predefined parameters, I simply do not know how to discover what predefined explicit parameters a definition is using from within a macro. I am not even sure if that can even be done at the macro level. The original `class abbrev` did not support it and since I lack the current technical expertise to know how to fix it, I just left it unresolved.  That seemed like wisest choice. Does that seem reasonable? 

The reason I can get around this for implicit and synthesized predefined parameters is because I can just let those parameters continue to be inferred/synthesize during application. However, for an explicit parameter, this is not an option. I somehow need to get the name of the applied variable (despite it not being present in the `class abbrev` definition) to use it at the application. This is the part I have no idea how to do.

leodemoura commented:
@tydeu Thanks for explaining the problem.
Yes, we currently cannot access the section variables from `MacroM`. 
So, to fix this limitation, we would have to write an elaboration function using `CommantElabM` which is inconvenient.
I think this is going to be an issue for other macros too.
@kha I think we should expose the section parameters at `MacroM`. Do you see other solutions?

tydeu commented:
I just pushed an commit that addresses the feedback. However, it could still probably use more tests. I also discovered and fixed an issue that resulted in no longer needing the complex `paramsToVars` helper, but rather just a simpler `paramsToArgs` helper. However, I am curious which of the following definitions fits better with Lean's code style.

This definition, which is more verbose but provides better error handling:

```lean
def paramsToArgs (params : Array Syntax) : MacroM (Array Syntax) := do
  let mut args := #[]
  for param in params do
    match param with
    | `(bracketedBinder| ( $ids:ident* $[: $type:term]? ) ) =>
      args := args ++ ids
    | `(bracketedBinder| [ $[ $id:ident : ]? $type:term ] ) =>
      continue
    | `(bracketedBinder| { $ids:ident* $[: $type ]? } ) =>
      continue
    | _ => 
      Lean.Macro.throwErrorAt param "unknown binder"
  return args
```

or something like the following, which is simpler but does not check for edge-case errors:

```lean
def paramsToArgs (params : Array Syntax) : MacroM (Array Syntax) := do
  let mut args := #[]
  for param in params do
    match param with
    | `(bracketedBinder| ( $ids:ident* $[: $type:term]? ) ) =>
      args := args ++ ids
    | _ => 
      continue
  return args
```

I went with the first one, but I can definitely see a strong case for simplifying it (especially as the "unknown binder" case should really never occur).

Kha reviewed (commented):
Thank you for your contribution. We have discussed the section variables issue internally and unfortunately it cannot be solved in a macro: even if we had access to the section variables, we need elaboration to (robustly) detect which section variables are actually used in the declaration. Rewriting the macro as an elaborator would be a big undertaking. Additionally, there are some concerns about performance with this kind of typeclass instance. So until we find time to redesign and reimplement this feature from the ground up, we would like to freeze the current restricted implementation. Do feel free to copy your extended version into a package of your own though and further experiment with it there though.

tydeu commented:
@Kha 

> We have discussed the section variables issue internally and unfortunately it cannot be solved in a macro: even if we had access to the section variables, we need elaboration to (robustly) detect which section variables are actually used in the declaration. Rewriting the macro as an elaborator would be a big undertaking. 

I personally don't think it should be an elaborator. I think having it not work with explicit section variables is a reasonable limitation.

> Additionally, there are some concerns about performance with this kind of typeclass instance. 

I am curious about what kind of performance issues these kind of typeclasses have, because I am making heavy use of this style of typeclass instance in one of my current projects. I would think it would be rather common pattern to want to abbreviate a  more general type classes with multiple more specific type classes.

> So until we find time to redesign and reimplement this feature from the ground up, we would like to freeze the current restricted implementation.

I would argue this change lifts a number of restrictions of the original version with no additional overhead (as far as I can see). In fact, this version seems to be simpler than the existing one, so I cannot see what the harm in it is.

However, I respect the burdens that come changing things and understand the desire to leave it as. Regardless, thank you for taking time to consider this change.

Kha commented:
Thanks for your understanding. It is true that you are lifting some restrictions without increasing the amount of code, but if that makes it works with only *some* section variables, that is not a definite plus. This command is simply so low priority for us compared to all the big open issues that we would like to not think about it at all right now. Similar with the performance issue, we have observed a real overhead in practice, but didn't get to actually investigating it yet. It could be that the overhead is not from the instance itself but that we simply replaced more specific constraints with a broader class abbrev.
